### ISSUE TITLE
Atypical "Erase In Display" behaviour

### ISSUE THREAD

**Issue by pfitzseb** [xtermjs/xterm.js#1727]
Most terminals seem to implement ED (specifically `\x1b[2J`) by not actually deleting anything from the display buffer, but rather by scrolling. Conversely, xterm.js does delete the display buffer's contents, which is a) inconsistent and b) annoying because you lose output. I don't think this behaviour qualifies as a bug because the spec doesn't say about what should happen with the scrollback buffer, but would still be nice if this were changed.

Downstream issue in https://github.com/JunoLab/atom-julia-client/issues/521.

## Details
- Browser and browser version: Electron 2.0.8
- OS version: Manjaro 
- xterm.js version: master (but also at least 3.6 and 3.7)

### Steps to reproduce
Try 
```bash
for i in {1..20}; do
  echo "$i"
done
echo '\e[1;1H\e[2J'
```
in xterm.js and another terminal emulator, and scroll up afterwards.

![image](https://user-images.githubusercontent.com/6735977/46527796-721bb900-c892-11e8-833d-58c15ef00477.png)
Expected behaviour is what e.g. Konsole does (middle window), and not how e.g. VSCode (left) or Juno's terminal (right, with xterm.js master) handles this.

**Comment by Tyriar** [comment]
Sounds good, Terminal.app and iterm2 also show this behavior.

**Comment by egmontkob** [comment]
[VTE](https://bugzilla.gnome.org/show_bug.cgi?id=506438) and [Konsole](https://bugs.kde.org/show_bug.cgi?id=384218) also behave as suggested here, and have bugreports to revert to xterm's behavior of deleting those lines :D

The main problem is: if someone's intent is really to clear the entire scrollback (which the `clear` command supposedly does), it no longer works completely, leaves a screenful of content in the scrollback because `clear` emits the sequences in the wrong order – well, at least wrong order for these kinds of terminal emulators.

It's either a downside here or a downside there, you pick :)

**Comment by Tyriar** [comment]
@egmontkob I thought clear typically sent a different sequence to clear the scrollback as well?

**Comment by seeruk** [comment]
Really keen to see this one resolved, it's the only thing holding me back from using terminal emulators like Hyper. I use Ctrl+L a lot to get rid of noise on my screen, but often want to be able to go back and see what was output. That includes things like testing. Personally, I don't see any reason not to work like iTerm2, VTE terminals, Terminal.app, etc. - there could be a separate way to clear the whole scrollback, or just push what's on screen up.

**Comment by egmontkob** [comment]
FYI: see also the latest comment in the aforelinked VTE thread – luckily the entire situation improved a lot by ncurses changing the order of the escape sequences it emits. If you decide to scroll out the lines, the `clear -x` command or bash's Ctrl+L will not leave yet another screenful of data in the scrollback.

**Comment by seeruk** [comment]
There is some good discussion in that other issue; are there plans to change this behaviour so that the screen is pushed up instead of erased?

**Comment by pjlsergeant** [comment]
Reading the history here, it seems a **_decision_** is needed before any more code is needed, and that is: what's the behaviour that this project wants to support?

The behaviour of `clear` is probably a red-herring, as its behaviour appears to be different on different systems: [this Linux example](https://linuxcommandlibrary.com/man/clear) (by default) clears scrollback too, where the OS X version does not (and doesn't claim to either). The bugs mentioned in https://github.com/xtermjs/xterm.js/issues/1727#issuecomment-427451617 appear to describe an issue with adding _empty lines_ to the scrollback.

@Tyriar Please could you make a decision either way? :-) The bug reports that keep getting duped suggest for many users the  _desired_ behaviour here is 2J shouldn't destroy any scrollback (and should neither add extra newlines to scrollback), but currently does, and that 3J should act the same way that 2J is currently handled?

**Comment by jerch** [comment]
Referring to the behavior of `clear` is indeed misleading because of the "E3" capability on some platforms.

Things get more clear, when looking at the sequences behind - ED2 and ED3:
- ED2 (CSI 2 J) - Erase all in activate display (cursor accessible region). There are two behavioral pattern - the "literal erase" one as seen in older devices and xterm doing a real erase (no scrollback additions), and a scroll-out one until the active region holds no data anymore. xterm.js implements the first one, which is more suitable in conjunction with ED3.
- ED3 (CSI 3 J) - Clear scrollback. Thats an xterm addition to the sequence realms. It is somewhat a surprising sequence, since it can directly manipulate scrollback, while kinda all other terminal content sequences are bound to the active cursor region. It is the "E3" capability denoted by some `clear` implementation. Support of the sequence varies across emulators, therefore `clear` tries to figured out by a terminfo lookup (thus TERM var dependent, quite often the source of issues here). xterm.js implements this feature.

@pjlsergeant Not sure what behavior you are looking for, but imho the pattern is obvious - we dont add things to scrollback on ED2 doing a real erase, and scrollback can be cleared with ED3. Ofc we could think about doing the scroll-out on ED2, but as mentioned by @egmontkob above this gets nasty in conjunction with ED3 support. Imho things work as intended - what do you think needs to be decided here?

**Comment by pjlsergeant** [comment]
@jerch If you think things work as intended, then presumably you are also suggesting this bug gets closed as WONTFIX?

**Comment by jerch** [comment]
Just looked at the ncurses changelog - actually the bug @egmontkob described above got fixed in `clear`. Furthermore for older versions still doing it in the wrong order it can be fixed easily with an alias as `clear && printf '\x1b[3J'`.

So I am open for the discussion, whether ED2 should erase or move things into the scrollbuffer. Personally I think scrolling is more useful for apps/users. Well such a mode needs a bit thinking about the details, how to scroll - Scroll up to active cursor position, erase below? Or move all lines up, or only up to the first line without any data? Well that sort of things...

**Comment by pjlsergeant** [comment]
>  I think scrolling is more useful for apps/users

Yes, and also seems to be common behaviour. 

> Well such a mode needs a bit thinking about the details

Sure, although again, I think all mentions of `clear` are a bit of a goose-chase, as the specific issue here is the handling of 2J, regardless of whatever `clear` is printing on a given day/system/etc. The behaviour of the following one-liner is easier to reason about:

`perl -e 'for ( 1 .. 100 ) { print "$_\n" }; print "\e[H\e[2J"'`

I was half-thinking of attempting a patch for this to make it work like _iTerm2_ does, but it doesn't seem like that a decision has been made about that being the desired behaviour, yet!

**Comment by jerch** [comment]
Reading through the bugtracker comments in vte and konsole raises alot of questions regarding scrolling out, and make me think whether this is a good idea at all:

ED2 was never meant to do that, its purpose is to erase content on the active viewport in the sense of "clean that area", nothing more. In fact apps might use it do full screen refreshs to start over drawing. I use it myself that way in several fullscreen scripts. Now if such an app does not switch to alternate buffer, it will dump every screen refresh into the scrollbuffer. The semantics are broken further by the fact, that ED1 does not add anything to the scrollbuffer. But why? By the spec description ED0 + ED1 should be equivalent to ED2, but it isnt anymore for scroll-out terminals. So semantically adding changes to the scrollbuffer for ED was a mistake, no clue who broke with the spec initially. Furthermore with SU (scroll up, `CSI <rows> S`) we already have a appropriate sequence for it. No need to water other sequences.

Now regarding whether this is common or not. I see that all OSX terminals I know do that, even Terminal.app. Under POSIX terminal emulators in general it is not that common - xterm (the father of all xterm-alikes and thus most common graphical TEs) doesnt do it, neither do the kernel consoles (tested linux and FreeBSD). The latter are more on the legacy side of things, but considering the semantic issues above, I am not so sure anymore if coupling scrollbuffer chnages to ED2 was a great idea.

**TL;DR**
After reading throught the bugtracker and concerns listed there, I lean towards not implementing scrollout for ED2. It breaks horribly the sequence semantics and the basic unix principle "do one thing and do it right". If ppl want scroll out, use the appropriate sequence for it (SU). All imho and up for discussion.

**Comment by pjlsergeant** [comment]
For many people, `xterm.js` will be the only terminal they use that doesn't do it "properly". Could this behaviour be behind a flag / OS check?

**Comment by jerch** [comment]
@pjlsergeant Yes, I am fine with doing it behind a custom settings, so users/integrators can decide, if they want it or not (maybe something like `scrollOnDisplayErase = true | false`). About the inner mechanics I'd prolly go with full page dumps with trimming empty lines from the bottom (iTerm2 seems to do it that way). Idk if any emulator respects scroll region settings / origin mode here, needs some testing (my guess is that none limits the scrolling by those settings).

Feel free to do a PR, imho we can discuss the inner mechanics over there.

**Comment by Tyriar** [comment]
New opt-in option coming in https://github.com/xtermjs/xterm.js/pull/5224 that allows scrolling instead of erasing the viewport
